Add "KubeServices" and "Nodes" to "ListResources" gRPC call

gabrielcorado opened the pull request:
Adds `KubeService` and `Node` to `ListResources` gRPC call. This PR is a continuation of #9096.

gabrielcorado committed:
feat: add KubeService and Node to ListResources

smallinsky commented on api/client/client.go:
```diff
@@ -823,15 +823,36 @@ func (c *Client) UpsertKubeService(ctx context.Context, s types.Server) error {
+	for i, resource := range resources {
+		srv, ok := resource.(types.Server)
+		if !ok {
+			return nil, trace.BadParameter("expected AppServer resource, got %T", resource)
```
```suggestion
			return nil, trace.BadParameter("expected Server  resource, got %T", resource)
```

gabrielcorado commented on lib/auth/auth_with_roles.go:
```diff
@@ -3140,30 +3190,43 @@ func (a *ServerWithRoles) GetKubeServices(ctx context.Context) ([]types.Server,
+
+// checkAccessToKubeService filters out kube clusters on each server and only
+// return the kube cluster the caller has access to.
+func (a *ServerWithRoles) checkAccessToKubeService(server types.Server) error {
```
Agreed, I've renamed it to `filterKubeServer` (similar to the one we already have for Nodes). Also, I added a comment on the function call to clarify that this flow works differently than the regular `checkAccessTo...`. Let me know if it makes more sense now.

Joerger commented on api/client/client.go:
```diff
@@ -823,15 +823,36 @@ func (c *Client) UpsertKubeService(ctx context.Context, s types.Server) error {
+	resources, err := c.GetResources(ctx, defaults.Namespace, types.KindKubeService)
 	if err != nil {
+		// DELETE IN 10.0
+		if trace.IsNotImplemented(err) {
+			resp, err := c.grpc.GetKubeServices(ctx, &proto.GetKubeServicesRequest{}, c.callOpts...)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			servers := make([]types.Server, len(resp.GetServers()))
+			for i, server := range resp.GetServers() {
+				servers[i] = server
+			}
+
+			return servers, nil
+		}
```
nit: Can you do what you did for ListNodes and move the fallback to a separate method?
```suggestion
		if trace.IsNotImplemented(err) {
			return c.getKubeServicesFallback(ctx)
		}
```

kimlisa commented:
as a continuation, could we also perhaps add `desktops` to `ListResources` as well?

r0mant commented on api/client/proto/authservice.proto:
```diff
@@ -1368,9 +1372,14 @@ message PaginatedResource {
-        types.DatabaseServerV3 DatabaseServer = 1;
-        // ApplicationServer represents a AppServer resource.
-        types.AppServerV3 AppServer = 2;
+        types.DatabaseServerV3 DatabaseServer = 1
+            [ (gogoproto.jsontag) = "database_server,omitempty" ];
+        // AppServer represents a AppServer resource.
+        types.AppServerV3 AppServer = 2 [ (gogoproto.jsontag) = "app_server,omitempty" ];
```
I'm not sure we can set jsontags here now since it wouldn't be a backwards compatible change.

gabrielcorado replied:
True, I will keep the fields only present at the new ones.

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -714,6 +714,40 @@ NextNode:
 	return filteredNodes, nil
 }
 
+func (a *ServerWithRoles) checkAccessToNode(server types.Server) error {
```
Seems like this function is very similar to `filterNodes`. Any way to reuse one or the other?

gabrielcorado replied:
Updated the `filterNodes` to reuse `checkAccessToNode`.

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -833,12 +869,23 @@ func (a *ServerWithRoles) checkAccessToResource(resource types.Resource) error {
+			return a.filterKubeServices(r)
+		}
+
+		return a.checkAccessToNode(r)
```
Since this returns an AccessDenied error, what would the behavior be if the user doesn't have access to any nodes? Will they get an access denied error or an empty list?

gabrielcorado replied:
They will receive and empty. This works the same way for all resources that the `ListResources` call can return.

r0mant commented on lib/services/local/presence_test.go:
```diff
@@ -587,8 +666,25 @@ func TestListResources(t *testing.T) {
 				resultResources = append(resultResources, resources...)
-			}
+				return len(resultResources) == totalResources
+			}, time.Second, time.Millisecond)
```
Seems quite aggressive to retry every millisecond even for tests. Same below.

fspmarshall reviewed (commented):
Couple security/perf tweaks needed:

- User-provided queries (i.e. label matching) need to be applied after role-based filtering.  We don't want users to be able to construct queries that are evaluated against resources that they don't have access to.  Even if timing attacks aren't currently possible (I wouldn't bet on that), it leaves the door open for future attacks to easily slip in if we establish a pattern of allowing user-supplied filters to circumvent RBAC.  Take a look at how `ListNodes` handles this for a good example.
- The way in which `ListNodes` falls back to serving pages from a ttl-cached copy of the complete node list when the cache is unhealthy is really important.  That behavior must be preserved (not falling back to the TTL-cache when the cache is unhealthy can basically brick dynamo clusters as it causes an infinite feedback-loop of throttling errors).  I'm not sure if the other resources supported by `ListResources` need this treatment right now, but they will at some point.

kimlisa commented:
I am not sure if we are matching against kube labels correctly. B/c the kube list we construct for the ui, we extract kube clusters from [each server](https://github.com/gravitational/teleport/blob/master/lib/web/ui/server.go#L112), then extract the labels. Shouldn't the label matching process be the same? Maybe i'm blind, or misunderstanding label matching for kube ;)

gabrielcorado commented:
@fspmarshall Both are very important points; thanks for raising them. About the first one, if I understood correctly, `ListNodes` matches labels (user input) [before](https://github.com/gravitational/teleport/blob/master/lib/services/local/presence.go#L293) and [after](https://github.com/gravitational/teleport/blob/master/lib/auth/auth_with_roles.go#L874) role-based filtering. Do you mean we should have both or only **after** filtering resources the user can access?

@kimlisa Those resources have multiple sets of labels. Based on the file you mentioned, we should use the labels at `KubernetesCluster` and `Database` instead of `Server` and `DatabaseServer`, respectively. We can keep the same labels as the file you mentioned for now, what do you think? Also, @r0mant, do you have any thoughts on this?

fspmarshall commented:
> `ListNodes` matches labels (user input) [before](https://github.com/gravitational/teleport/blob/master/lib/services/local/presence.go#L293) and [after](https://github.com/gravitational/teleport/blob/master/lib/auth/auth_with_roles.go#L874) role-based filtering.

@gabrielcorado Yeah, the logic is implemented in both places.  This is necessary because not all calls pass through the RBAC layer (e.g. if the call hits a local cache).  The RBAC layer pulls the label query out of the incoming request [here](https://github.com/gravitational/teleport/blob/86e92be2532958777561b61fe84b5604dbc7cb8b/lib/auth/auth_with_roles.go#L856-L858) so that they don't get applied until after RBAC filtering.

gabrielcorado commented:
@fspmarshall Thanks for your explanation. I'll update the `ListResources` and add comments about this.

kimlisa commented:
> Those resources have multiple sets of labels. Based on the file you mentioned, we should use the labels at KubernetesCluster and Database instead of Server and DatabaseServer, respectively. We can keep the same labels as the file you mentioned for now, what do you think? Also, @r0mant, do you have any thoughts on this?

I wasn't sure what you meant, but maybe this is what you meant: I noticed we have this for desktops, [GetWindowsDesktops](https://github.com/gravitational/teleport/blob/master/lib/auth/auth_with_roles.go#L3598) (`KindWindowsDesktop`) and [GetWindowsDesktopServices](https://github.com/gravitational/teleport/blob/master/lib/auth/auth_with_roles.go#L3562) (`KindWindowsDesktopService`). We can keep current behavior for kube, db, app servers as is and just add other resource kinds (ie: if we want list of kube clusters instead of kube services, we can use `KindKubernetesCluster`) to our `ListResource` switch statement and handle it from there. I can do these parts if you want.

gabrielcorado commented:
@fspmarshall I updated the `ListResources` in the RBAC layer [to perform the label matching](https://github.com/gravitational/teleport/pull/9613/files#diff-ec1bd6a036583633acf2fa781d0d32d4355c2f144ce484fa9a7f780b89c3311dR832). Also, I added the [cache unhealthiness fallback logic for nodes](https://github.com/gravitational/teleport/pull/9613/files#diff-3e237f973a30f7092dc18a706ecd95d8a7e1d99c5c50626b8bb0e7ac6cb7efd1R1760) as the other resources do not implement something like this (we can think of adding for them in future PRs). 

@kimlisa I updated the labels used for label matching. They are more aligned with the files you mentioned. [KubeService](https://github.com/gravitational/teleport/pull/9613/files#diff-cfa6ef287887d6818cff1f237d2ac06b6fc77e6ea82d993fe7ab7aaa42be4ae0R280), [DatabaseServer](https://github.com/gravitational/teleport/pull/9613/files#diff-e487bd45f3d7e8e05597facf82cbf12b5ee8339df7b0d28534ce01c4c7a47da2R259) and [AppServer](https://github.com/gravitational/teleport/pull/9613/files#diff-cd256606a0263c4f0caebae17bcd03a2d6a4066c9a5bb1a5656aa389e8cde89bR268) were updated.

Can you retake a look, please?

kimlisa commented on lib/auth/auth_with_roles.go:
```diff
@@ -783,30 +788,38 @@ func (a *ServerWithRoles) GetNodes(ctx context.Context, namespace string, opts .
-		resourceKind = types.KindDatabaseServer
-	case types.KindAppServer:
-		resourceKind = types.KindAppServer
+	case types.KindDatabaseServer,
+		types.KindAppServer,
+		types.KindNode,
+		types.KindKubeService:
```
Sorry gabriel, I don't think I was clear on my earlier comment. What I was suggesting was in addition to these kinds, we should also add `KindApplication` and `KindDatabase` that returns list of apps and dbs instead of servers. It will be similar to desktops. When it's time to add desktops here, we will need both `KindWindowsDesktopService` which is equiv to `GetWindowsDesktopService` and `KindWindowsDesktop` -> `GetWindowsDesktop`. I would leave label matching as it was before, unless we got confirmation that merging of labels is the behavior we want

fspmarshall commented on lib/auth/auth_with_roles.go:
```diff
@@ -833,12 +851,23 @@ func (a *ServerWithRoles) checkAccessToResource(resource types.Resource) error {
 	case types.DatabaseServer:
 		return a.checkAccessToDatabase(r.GetDatabase())
+	case types.Server:
+		if resource.GetKind() == types.KindKubeService {
+			// KubeServices are present inside the type Server. We need to
+			// filter out the ones that the user doesn't have access to. This
+			// function won't return an AccessDenied because we're not checking
+			// access to the Server itself.
+			return a.filterKubeServices(r)
+		}
+
+		return a.checkAccessToNode(r)
```
Since permissions can vary based on resource kind here, we should err out on unexpected resource kinds rather than assuming that the resource is a node if it isn't a kube service (help catch future bugs).

r0mant reviewed (approved):
@gabrielcorado Left a comment about including app's dynamic labels into GetAllLabels method but otherwise lgtm.
